### PR TITLE
Upgrade JDK to 21

### PR DESCRIPTION
**Pull request by quantranhong1999**
continuation of https://github.com/apache/james-project/pull/1939

### CODE REVIEW

**Comment by quantranhong1999** [comment]
The failed webadmin tests, I guess it is related to current master issue.

`JPARecomputeCurrentQuotasServiceTest` and `MemoryRecomputeCurrentQuotasServiceTest` failures seem related though.

**Comment by Arsnael** [comment]
I'm trying to get a green build for master but I think the commits here should do the trick: https://github.com/apache/james-project/pull/1957 

Feel free to cherry-pick them

**Comment by quantranhong1999** [comment]
Green.

Some unstable tests failing though: https://ci-builds.apache.org/blue/organizations/jenkins/james%2FApacheJames/detail/PR-1963/2/tests/

```java
Unstable Tests / (?) – org.apache.james.rrt.cassandra.RewriteTablesTest
<1s
Error
Suite [org.apache.james.rrt.cassandra.RewriteTablesTest] did not discover any tests
Stacktrace
org.junit.platform.suite.engine.NoTestsDiscoveredException: Suite [org.apache.james.rrt.cassandra.RewriteTablesTest] did not discover any tests
```

Maybe related to cucumber changes. I would check this.

But not a blocker IMO.

**Comment by chibenwa** [comment]
Please take the time to investigate...

**Comment by chibenwa** [comment]
> so we can fasten the JDK21-related

Why need to fasten? Likely a vote on the mailing list would be a pre-requisite to a merge...

> fix the broken master also)

Let's merge https://github.com/apache/james-project/pull/1957 instead to achieve this goal.

**Comment by quantranhong1999** [comment]
The LDAP test failures seems not related to me (green locally).

**Review by Arsnael — APPROVED** [review]

**Review by vttranlina — APPROVED** [review]

File: mailbox/jpa/pom.xml [hunk 1]
```diff
@@ -31,6 +31,10 @@
     <packaging>jar</packaging>
     <name>Apache James :: Mailbox :: JPA</name>
 
+    <properties>
```
Comment (jeantil) [inline mailbox/jpa/pom.xml:34]: How does this behave in IDEs ? does it mean we need to have both jdk 11 and jdk21 installed to be able to build james ?

File: pom.xml [hunk 2]
```diff
@@ -3514,7 +3514,7 @@
                         <configuration>
                             <rules>
                                 <requireMavenVersion>
-                                    <version>3.6.0</version>
+                                    <version>3.8.1</version>
```
Comment (jeantil) [inline pom.xml:3549]: why not up to the latest (3.9.6) ?

File: pom.xml [hunk 3]
```diff
@@ -2962,6 +2962,21 @@
...
 
+    <pluginRepositories>
+        <pluginRepository>
+            <releases>
+                <enabled>false</enabled>
+            </releases>
+            <snapshots>
+                <enabled>true</enabled>
+            </snapshots>
+            <id>apache.snapshots</id>
```
Comment (jeantil) [inline pom.xml:2983]: this doesn't seem related to upgrading mockito ... and I may be naive but we pull snapshot versions of plugins ? 
doesn't that make our build non reproductible since snapshot versions may be updated completely randomly ?

File: mailbox/jpa/pom.xml [hunk 4]
```diff
@@ -91,6 +91,7 @@
         <dependency>
             <groupId>${james.groupId}</groupId>
             <artifactId>james-json</artifactId>
+            <type>test-jar</type>
```
Comment (jeantil) [inline mailbox/jpa/pom.xml:94]: fixup! 2e1afc4281c23d48d53c3ffc015d1594fe06de89

File: mailbox/memory/pom.xml [hunk 5]
```diff
@@ -72,6 +72,7 @@
         <dependency>
             <groupId>${james.groupId}</groupId>
             <artifactId>james-json</artifactId>
+            <type>test-jar</type>
```
Comment (jeantil) [inline mailbox/memory/pom.xml:75]: fixup! 2e1afc4281c23d48d53c3ffc015d1594fe06de89 JAMES-3961 AssertJ bean introspection

File: pom.xml [hunk 6]
```diff
@@ -621,7 +621,7 @@
         <concurrent.version>1.3.4</concurrent.version>
         <xbean-spring.version>4.21</xbean-spring.version>
         <netty.version>4.1.101.Final</netty.version>
-        <cucumber.version>7.14.1</cucumber.version>
+        <cucumber.version>7.15.0</cucumber.version>
```
Comment (jeantil) [inline pom.xml:624]: seems like a fixup!  JAMES-3691 Update Cucumber to 7.14.1 

i wouldn't mind a longer explanation of why rrt tests needed to be dropped

File: pom.xml [hunk 7]
```diff
@@ -2831,11 +2831,6 @@
                 <artifactId>mockito-core</artifactId>
                 <version>${mockito.version}</version>
             </dependency>
-            <dependency>
```
Comment (jeantil) [inline pom.xml:2819]: hmm maybe fixup! 
JAMES-3961 Upgrade mockito: 5.2.0 -> 5.7.0

**Comment by jeantil** [comment]
I don't understand commits 
-  Revert "[UPDATE] Update spark version: Jetty update" 
-  Revert "[UPDATE] BouncyCastle 1.70 -> 1.77 (fixes multiple minor CVEs)" 
well spark maybe, I saw in the other PR that it was not a great upgrade, but reverting security patches really sounds weird

File: pom.xml (hunk 3 above)
Comment (chibenwa) [reply pom.xml:2983]: Yes related to some tries on JPA updades +1 shall be removed...

**Comment by chibenwa** [comment]
> I don't understand commits

>>    Revert "[UPDATE] Update spark version: Jetty update"
>>    Revert "[UPDATE] BouncyCastle 1.70 -> 1.77 (fixes multiple minor CVEs)"
>>    well spark maybe, I saw in the other PR that it was not a great upgrade, but reverting security patches really sounds weird

Those were to fix the build at a previous point-in-time. This would be removed by a rebase.

**Comment by Arsnael** [comment]
>   well spark maybe, I saw in the other PR that it was not a great upgrade, but reverting security patches really sounds weird

As said by @chibenwa, need a good rebase. Those reverts are present in the code to fix the build, cause I accidentally merged them, as their respective PRs were green but unfortunately did not realize that Github seems to think it's ok if CI is unavailable and you have no build linked to the PR (yes weird... but my bad). Thus the reverts.

FYI: 
- The bouncycastle upgrade has been repushed and fixed and merged: https://github.com/apache/james-project/pull/1975
- the spark upgrade was an attempt based on a fork as the original lib seems unmaintained for some time, but it got abandoned for now as it brings more issues than anything else for now: https://github.com/apache/james-project/pull/1974

File: mailbox/jpa/pom.xml (hunk 1 above)
Comment (quantranhong1999) [reply mailbox/jpa/pom.xml:34]: > does it mean we need to have both jdk 11 and jdk21 

I believe we only need jdk21 installed.
I set IDE and terminal jdk to jdk21 only. Then IDE build and terminal build are both fine.

File: pom.xml (hunk 2 above)
Comment (quantranhong1999) [reply pom.xml:3549]: can try on that

File: pom.xml (hunk 6 above)
Comment (quantranhong1999) [reply pom.xml:624]: > i wouldn't mind a longer explanation of why rrt tests needed to be dropped

Do you mean the data-file RRT test removal? (https://github.com/apache/james-project/pull/1963/commits/ccce958046ec774097b6eb73329a48de97c62e92#diff-c7eba70ed09589fac26f92d4756f4e266aa408f64337f17131f77f1b1333bb13)
image

cc @chibenwa any rationale behind that?

File: pom.xml (hunk 2 above)
Comment (quantranhong1999) [reply pom.xml:3549]: Anyway, it is likely just a minimal Maven version to support Jdk21, no special reason behind that. Can just work well with the latest 3.9.6 I guess.

Comment (jeantil) [reply pom.xml:3549]: ok thanks for the explanation

File: pom.xml (hunk 6 above)
Comment (jeantil) [reply pom.xml:624]: the commit comment only says a cryptic 
> RRT cucumber tests do not account for read-only 
what seems weird is that these tests have been aroun for a while and I don't understand why moving to jdk21 would affect them. 
Does it mean that moving to jdk21 affects : 
- the capabilities of the data-file rrt ? 
- only the tests and we decided to just drop the tests (seems weird)
- was this already broken somehow before the upgrade and is actually a cleanup commit that's unrelated to jdk21

**Review by hungphan227 — APPROVED** [review]

File: pom.xml (hunk 6 above)
Comment (quantranhong1999) [reply pom.xml:624]: > only the tests and we decided to just drop the tests (seems weird)

It seems to fall in this case.

The test was still fine before this work.

RRT tests for data-file is just only 1 test (with `readonly` tag): https://github.com/apache/james-project/blob/master/server/data/data-library/src/test/resources/cucumber/rewrite_tables.feature#L4 though.

I will try to see if we can still keep that, otherwise only 1 test lost is minimal and could be acceptable IMO.

**Comment by quantranhong1999** [comment]
CC @jeantil I reworked a bit the commit history and brought back the RRT data-file test.

**Comment by jeantil** [comment]
thanks I'll try and review again over lunch

File: server/container/lifecycle-api/src/test/java/org/apache/james/lifecycle/api/LeakAwareTest.java [hunk 8]
```diff
@@ -76,18 +73,13 @@ public static ListAppender<ILoggingEvent> getListAppenderForClass(Class clazz) {
         return loggingEventListAppender;
     }
 
-    private void forceChangeLevel(String level) throws NoSuchFieldException, IllegalAccessException {
-        forceChangeLevel(LeakAware.Level.parse(level));
+    private void forceChangeLevel(String level) {
+        LeakAware.LEVEL = LeakAware.Level.parse(level);
     }
 
     // using reflect to change LeakAware.LEVEL value
```
Comment (jeantil) [inline server/container/lifecycle-api/src/test/java/org/apache/james/lifecycle/api/LeakAwareTest.java:80]: I don't think the comment applies anymore :D

**Review by jeantil — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Rebase needed

**Comment by quantranhong1999** [comment]
There was a minor conflict in pom.xml, so I rebased and force pushed.

**Comment by quantranhong1999** [comment]
Green :-)

**Comment by chibenwa** [comment]
Let's merge?

**Comment by quantranhong1999** [comment]
> Let's merge?

We could I guess. Don't forget to conclude the decision on the mailing list ^^.
